source/extensions/io_socket/io_uring/io_handle_impl.cc

```diff
@@ -0,0 +1,357 @@
+#include "source/extensions/io_socket/io_uring/io_handle_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/event/dispatcher.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/network/address_impl.h"
+#include "source/common/network/io_socket_error_impl.h"
+#include "source/extensions/io_socket/io_uring/io_uring.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace IoSocket {
+namespace IoUring {
+
+namespace {
+
+constexpr socklen_t udsAddressLength() { return sizeof(sa_family_t); }
+
+} // namespace
+
+IoUringSocketHandleImpl::IoUringSocketHandleImpl(const uint32_t read_buffer_size,
+                                                 const IoUringFactory& io_uring_factory, os_fd_t fd,
+                                                 bool socket_v6only, absl::optional<int> domain)
+    : read_buffer_size_(read_buffer_size), io_uring_factory_(io_uring_factory), fd_(fd),
+      socket_v6only_(socket_v6only), domain_(domain) {}
+
+IoUringSocketHandleImpl::~IoUringSocketHandleImpl() {
+  if (SOCKET_VALID(fd_)) {
+    IoUringSocketHandleImpl::close();
+  }
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::close() {
+  ASSERT(SOCKET_VALID(fd_));
+  io_uring_factory_.getOrCreateUring().prepareClose(fd_);
+  if (isLeader()) {
+    io_uring_factory_.getOrCreateUring().unregisterEventfd();
+    file_event_adapter_.reset();
+  }
+  SET_SOCKET_INVALID(fd_);
+  return Api::ioCallUint64ResultNoError();
+}
+
+bool IoUringSocketHandleImpl::isOpen() const { return SOCKET_VALID(fd_); }
+Api::IoCallUint64Result IoUringSocketHandleImpl::readv(uint64_t /* max_length */,
+                                                       Buffer::RawSlice* /* slices */,
+                                                       uint64_t /* num_slice */) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                      absl::optional<uint64_t> max_length_opt) {
+  const uint64_t max_length = max_length_opt.value_or(UINT64_MAX);
+  if (max_length == 0) {
+    return Api::ioCallUint64ResultNoError();
+  }
+
+  if (bytes_to_read_ == 0) {
+    return Api::IoCallUint64Result(
+        0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                           Network::IoSocketError::deleteIoError));
+  }
+
+  ASSERT(read_buf_ != nullptr);
+  auto fragment = new Buffer::BufferFragmentImpl(
+      read_buf_.get(), bytes_to_read_,
+      [](const void* /*data*/, size_t /*len*/, const Buffer::BufferFragmentImpl* this_fragment) {
+        delete this_fragment;
```

Comment:
What happens if the BufferFragmentImpl in buffer out-lives the IoUringSocketHandleImpl that owns read_buf_ ?  Seems like crash potential.

Reply:
True. Now I transfer the ownership of read_buf_ to the BufferFragmentImpl and delete it in the releaser.